settings: new notification API

RaduBerinde opened the pull request:
This is a proposal for a new API for getting notifications for changes to cluster settings.

As discussed in more detail in #73830, the existing mechanism is fragile: it executes an arbitrary callback inside the updater, possibly blocking it.

The new API is to create a `Notifier` which has a buffer channel that the caller can listen on. The `Notifier` is associated with a set of settings, and a non-blocking send is performed on the channel whenever one of the settings changes.

Release note: None
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98280)
<!-- Reviewable:end -->

dt commented on pkg/settings/notifer.go:
```diff
@@ -0,0 +1,76 @@
+	}
+	if buildutil.CrdbTestBuild {
+		n.debugStack = string(debug.Stack())
+		runtime.SetFinalizer(n, func(obj interface{}) {
```
We could just return a close func() to make it obvious you should defer it?

dt commented on pkg/settings/notifer.go:
```diff
@@ -0,0 +1,76 @@
+	for i := range settings {
+		slots[i] = settings[i].slotIdx()
+	}
+	ch := make(chan struct{}, 1)
```
I wonder if we should let the caller make their own ch to choose the buffering capacity (e.g. if they expect to be blocked more or something)? 

Would it make sense to just have `func (v Values) Subscribe(ch chan<- struct{}, settings...) (unsubscribe func())` ?

dt replied:
oh, nevermind, I see that's exactly what Andrew proposed in https://github.com/cockroachdb/cockroach/issues/73830#issuecomment-995007403 and you already said you preferred this instead.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98280)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @knz)

___
*[`pkg/settings/notifer.go` line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-NQ3M6M2LVXuZxTHPARz-r1-32:-NQ6VrdZC4p1ozbwZBaI:bffuld5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c19496d72b638f632efe15e039afa790ddaa1f02/pkg/settings/notifer.go#L32)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

oh, nevermind, I see that's exactly what Andrew proposed in https://github.com/cockroachdb/cockroach/issues/73830#issuecomment-995007403 and you already said you preferred this instead.
</blockquote></details>

Yeah, I can't imagine a usecase where you would need a bigger buffer - why would you care how many changes happened before you got a chance to read from the channel? You have to re-read the current values at that point anyway.
___
*[`pkg/settings/notifer.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-NQ3M6M2LVXuZxTHPARz-r1-41:-NQ6WD885Gr2YVaJICmk:b-tvodzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c19496d72b638f632efe15e039afa790ddaa1f02/pkg/settings/notifer.go#L41)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

We could just return a close func() to make it obvious you should defer it?
</blockquote></details>

I wasn't quite sure if the typical usecase would involve `defer` or would involve storing a `*Notifier` and making it sure it gets closed (e.g. by the stopper). In any case, I prefer having a type with documentation instead of just returning a channel and a func (there would be no easy way from a code navigation perspective to get from a use of that channel to any of the documentation here). I think having to `Close` things is a pretty established pattern so I don't see the advantage of a bare func()


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
LGTM with nits

> Reviewed 3 of 6 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-:-NQZeIltDvS0cfXtkQt2:b3nilp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @RaduBerinde)

___
*[`pkg/settings/values.go` line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-NQZeE8T5rSztwisGEOM:-NQZeE8T5rSztwisGEON:b-r84rzj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ab88926c172fc9980658b84c9eb5ffb3e716eb4/pkg/settings/values.go#L213)):*
> ```Go
> 		sv.changeMu.onChangeCh[slot] = append(sv.changeMu.onChangeCh[slot], ch)
> 	}
> 	sv.changeMu.Unlock()
> ```

defer this
___
*[`pkg/settings/values.go` line 228 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-NQZeF7H54sJRuYdgpgk:-NQZeF7H54sJRuYdgpgl:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ab88926c172fc9980658b84c9eb5ffb3e716eb4/pkg/settings/values.go#L228)):*
> ```Go
> 		}
> 	}
> 	sv.changeMu.Unlock()
> ```

ditto


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
settings: new notification API

This is a proposal for a new API for getting notifications for changes
to cluster settings.

As discussed in more detail in #73830, the existing mechanism is
fragile: it executes an arbitrary callback inside the updater,
possibly blocking it.

The new API is to create a `Notifier` which has a buffer channel that
the caller can listen on. The `Notifier` is associated with a set of
settings, and a non-blocking send is performed on the channel
whenever one of the settings changes.

Release note: None
Epic: none

RaduBerinde reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-:-NR529sn4cJUUkghTJhu:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @knz)

___
*[`pkg/settings/values.go` line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-NQZeE8T5rSztwisGEOM:-NR528uD70RInWvkYOr3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ab88926c172fc9980658b84c9eb5ffb3e716eb4/pkg/settings/values.go#L213)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

defer this
</blockquote></details>

Done.
___
*[`pkg/settings/values.go` line 228 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98280#-NQZeF7H54sJRuYdgpgk:-NR529ID0-q7f2BBytTq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ab88926c172fc9980658b84c9eb5ffb3e716eb4/pkg/settings/values.go#L228)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
The existing API has a big advantage when it comes to testing - when you override a setting, you are guaranteed that the on-change callback ran by the time the `Override()` call returns. There are tests which call `Override()` and expect the change to have taken effect immediately. I don't quite know how to achieve something similar. Even if we have a facility that waits for all notifier channels to be "drained",  there is no guarantee that the relevant code finishes running.
